ghproxy/ghmetrics/ghpaths.go

```diff
@@ -0,0 +1,295 @@
+	fragment := getFirstFragment(path)
+	switch fragment {
+	case "repos":
+		// /:owner/:repo
+		return handleRepos(path)
+	case "user":
+		// /user
+		return handleUser(path)
+	case "users":
+		// /users
+		return handleUsers(path)
+	case "orgs":
+		// /orgs
+		return handleOrgs(path)
+	case "issues":
+		// /issues
+		return fmt.Sprintf("/%s%s", fragment, handleConstantAndVar(path))
+	case "search":
+		// do we care to handle search sub-paths differently?
+		// e.g.: /search/repositories, /search/commits, /search/code, /search/issues, /search/users, /search/topics, /search/labels
+		return path
+	case "gists":
+		// do we care to handle gist sub-paths differently?
+		// e.g. /gists/public, /gists/starred
+		return path
+	case "notifications":
+		// /notifications
+		return handleNotifications(path)
+	case "repositories", "emojis", "events", "feeds", "hub", "rate_limit", "teams", "licenses":
+		return path
+	default:
+		logrus.WithField("path", path).Warning("Path not handled")
+		return path
+	}
+}
+
+// getFirstFragment returns the first fragment of a path, of a
+// `*url.URL.Path`. e.g. `/repos/kubernetes/test-infra/` will return `repos`
+func getFirstFragment(path string) string {
+	result := strings.Split(path, "/")
+	for _, str := range result {
+		if str != "" {
+			return str
+		}
+	}
+	return ""
+}
+
+func handleRepos(path string) string {
+	match := reposRegex.FindStringSubmatch(path)
+	result := make(map[string]string)
+	for i, name := range reposRegex.SubexpNames() {
+		if i != 0 && name != "" && i <= len(match) { // skip first and empty
+			if match[i] != "" {
+				result[name] = match[i]
+			}
+		}
+	}
+	if result["owner"] == "" || result["repo"] == "" {
+		logrus.WithField("path", path).Warning("Not handling /repos/.. path correctly")
+		return "/repos"
+	}
+	rest := result["rest"]
+	sanitizedPath := fmt.Sprintf("/repos/%s/%s", ":owner", ":repo")
+	if rest == "" || rest == "/" {
+		return sanitizedPath
+	}
+	switch fragment := getFirstFragment(rest); fragment {
+	case "issues", "branches":
+		return fmt.Sprintf("%s%s", sanitizedPath, handlePrefixedVarAndConstant(fmt.Sprintf("/%s", fragment), rest))
+	case "keys", "labels", "milestones", "pulls", "releases", "statuses", "subscribers", "assignees", "archive", "collaborators", "comments", "compare", "contents", "commits":
+		// archive is a special path that might need better handling
+		return fmt.Sprintf("%s/%s%s", sanitizedPath, fragment, handleConstantAndVar(rest))
+	case "git":
+		return fmt.Sprintf("%s%s", sanitizedPath, handlePrefixedConstantAndVar(fmt.Sprintf("/%s", fragment), rest))
+
+	case "merges", "stargazers", "notifications", "hooks":
+		return fmt.Sprintf("%s%s", sanitizedPath, rest)
+	default:
+		logrus.WithField("sanitizedPath", sanitizedPath).WithField("rest", rest).Warning("Path not handled")
+		return fmt.Sprintf("%s%s", sanitizedPath, rest)
+	}
+}
+
+func handleUser(path string) string {
+	match := userRegex.FindStringSubmatch(path)
```

Comment:
nit: could we move the associated regex into these functions instead of having them at the top? makes it harder t oread